File: src/foam/lib/parse/TrapPStream.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.lib.parse;

/**
 * This PStream is used to test against a single character to check if the parser accepts it or not
 */
public class TrapPStream
    implements PStream
{
  protected ErrorReportingPStream root;
  protected Character head;

  public TrapPStream(ErrorReportingPStream root) {
    this(root, null);
  }

  public TrapPStream(ErrorReportingPStream root, Character head) {
    this.root = root;
    this.head = head;
  }

  @Override
  public char head() {
    return this.head;
  }

  public void setHead(char head) {
    this.head = head;
  }

  @Override
  public boolean valid() {
    return true;
  }

  @Override
  public PStream tail() {
    root.reportValidCharacter(head);
    return InvalidPStream.instance();
  }

  @Override
  public Object value() {
    return null;
  }

  @Override
  public PStream setValue(Object value) {
    return this;
  }

  @Override
  public String substring(PStream end) {
    return null;
  }

  @Override
  public PStream apply(Parser parser, ParserContext x) {
    return parser.parse(this, x);
  }
}

File: src/foam/parse/TrapPStream.js
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.parse',
  name: 'TrapPStream',

  properties: [
    {
      name: 'delegate'
    },
    {
      name: 'head',
      getter: function() { return this.delegate.head; },
    },
    {
      name: 'tail',
      getter: function() {
        throw "trap";
        return foam.parse.InvalidPStream.create();
      },
      setter: function(){}
    },
    {
      name: 'valid',
      getter: function() { return this.delegate.valid; }
    },
    {
      name: 'value',
      getter: function() { return this.delegate.value; }
    }
  ],

  methods: [
    function initArgs() {},

    function setValue(value) {
      if ( value === undefined ) value = null;
      var ps = this.cls_.create();
      ps.delegate = this.delegate.setValue(value);
      ps.report = this.report;
      return ps;
    },

    function substring(end) {
      return null;
    },

    function apply(p, obj) {
      return p.parse(this, obj);
    },

    function compareTo(other) {
      return this.delegate.compareTo(other.delegate);
    }
  ]
});
